mono/mini/method-to-ir.c

```diff
@@ -4311,6 +4311,14 @@ mini_emit_array_store (MonoCompile *cfg, MonoClass *klass, MonoInst **sp, gboole
 		iargs [1] = sp [1];
 		iargs [0] = sp [0];
 
+		MonoClass *array_class = sp [0]->klass;
+		if (array_class && m_class_get_rank (array_class) == 1) {
+			MonoClass *eclass = m_class_get_element_class (array_class);
+			if (m_class_is_sealed (eclass))
+				/* Make a non-virtual call if possible */
+				return mono_emit_method_call (cfg, helper, iargs, NULL);
```

Comment:
the problem is that `mono_marshal_get_vritual_stelemref` call above is passed `obj_array` as the type, which makes it always choose the `STELEMREF_OBJECT` wrapper kind which has no typecheck at all.  So you need to pass a more precise static class.

(Seems like it should've already been broken for the virtual case.  I'm not sure.)